### PR TITLE
Don't derive FE::InternalDataBase from Mapping::InternalDataBase.

### LINKED ISSUE #1280: FEValues cleanup: Don't derive FiniteElement::InternalDataBase from Mapping::InternalDataBase
Right now, this is indeed the case although there is no real (apparent reason) for doing so other than the fact that finite elements can use at least a subset of the fields that the mapping also needs. Logically, it may be cleaner to just have these classes be separate and, as necessary, duplicate the member variables.

In fact, one could argue that the base classes should not have any member variables at all, since not all finite element (or mapping) classes need all of them. For example, all of the non-parametric elements don't need to store anything in there anyway. Rather, each element should decide which members it needs, rather than inheriting a grab bag from the base class.

This requires the resolution of #1279 first.

### PR DESCRIPTION
**Pull request by bangerth**
There was no logical reason to do this, and it turns out that the derived classes
actually didn't need any of the fields of the class it was previously derived from (other than the
is_first_cell/clear_first_cell mechanism, see #1305). So break the inheritance
and have these two classes be completely unrelated.

While there, also use unique_ptr rather than SmartPointer to point to such objects
in FEValues -- these pointers are never shared between objects, so ownership is
always clear. Doing so requires the addition of a MemoryConsumption function for
unique_ptr objects, but it allows us to not derive the InternalDataBase objects
from Subscriptor any more.

This fixes #1280. In reference to #1198.

### CODE REVIEW

**Comment by bangerth** [comment]
Passes the testsuite, with the exception of the currently failing BDM tests.

File: include/deal.II/base/memory_consumption.h [hunk 1]
```diff
@@ -631,6 +641,16 @@ namespace MemoryConsumption
   template <typename T>
   inline
   std::size_t
+  memory_consumption (const std_cxx11::unique_ptr<T> &)
+  {
+    return sizeof(std_cxx11::unique_ptr<T>);
```
Comment (kronbichler) [inline include/deal.II/base/memory_consumption.h]: Here I'm wondering whether you shouldn't also include the memory consumed of T. WIth a unique_ptr, it is clear that the memory is where the pointer is. (shared_ptr is a different story.) But I'll have to check below how you use it.

**Comment by kronbichler** [comment]
Apart from the issue with memory consumption of unique_ptr, the patch is fine for merging.

File: include/deal.II/base/memory_consumption.h (hunk 1 above)
Comment (bangerth) [inline include/deal.II/base/memory_consumption.h]: It's a fair question. I simply opted for consistency with the other pointer types.

The old code already added the memory of the pointer plus the memory of the object pointed to. I didn't change that in this patch (nor should it be changed).

Comment (kronbichler) [inline include/deal.II/base/memory_consumption.h]: I agree it's best to treat pointers the same way.

**Pull request merged**